cedar-policy-validator/src/entity_manifest.rs

```diff
@@ -0,0 +1,1103 @@
+    /// A entities error was encountered
+    #[error(transparent)]
+    #[diagnostic(transparent)]
+    EntitiesError(#[from] EntitiesError),
+
+    /// The request was partial
+    #[error("Entity slicing requires a fully concrete request. Got a partial request.")]
+    PartialRequestError,
+    /// A policy was partial
+    #[error(
+        "Entity slicing requires fully concrete policies. Got a policy with an unknown expression."
+    )]
+    PartialExpressionError,
+
+    /// A policy was not analyzable because it used unsupported operators
+    /// before a [`ExprKind::GetAttr`]
+    /// See [`FailedAnalysisError`] for more details.
+    #[error(transparent)]
+    #[diagnostic(transparent)]
+    FailedAnalysis(#[from] FailedAnalysisError),
+}
+
+/// Union two tries by combining the fields.
+fn union_fields<T: Clone>(first: &Fields<T>, second: &Fields<T>) -> Fields<T> {
+    let mut res = first.clone();
+    for (key, value) in second {
+        if let Some(existing) = res.get(key) {
+            res.insert(key.clone(), Box::new((*existing).union(value)));
+        } else {
+            res.insert(key.clone(), value.clone());
+        }
+    }
+    res
+}
+
+impl AccessPath {
+    /// Convert a [`AccessPath`] into corresponding [`RootAccessTrie`].
+    fn to_primary_slice(&self) -> RootAccessTrie {
+        self.to_primary_slice_with_leaf(AccessTrie {
+            parents_required: true,
+            children: Default::default(),
+            data: (),
+        })
+    }
+
+    /// Convert an [`AccessPath`] to a [`RootAccessTrie`], and also
+    /// add a full trie as the leaf at the end.
+    fn to_primary_slice_with_leaf(&self, leaf_trie: AccessTrie) -> RootAccessTrie {
+        let mut current = leaf_trie;
+        // reverse the path, visiting the last access first
+        for field in self.path.iter().rev() {
+            let mut fields = HashMap::new();
+            fields.insert(field.clone(), Box::new(current));
+            current = AccessTrie {
+                parents_required: false,
+                children: fields,
+                data: (),
+            };
+        }
+
+        let mut primary_map = HashMap::new();
+        primary_map.insert(self.root.clone(), current);
+        RootAccessTrie { trie: primary_map }
+    }
+}
+
+impl RootAccessTrie {
+    /// Create an empty [`RootAccessTrie`] that requests nothing.
+    pub fn new() -> Self {
+        Self {
+            trie: Default::default(),
+        }
+    }
+}
+
+impl<T: Clone> RootAccessTrie<T> {
+    /// Union two [`RootAccessTrie`]s together.
+    /// The new trie requests the data from both of the original.
+    fn union(&self, other: &Self) -> Self {
+        let mut res = self.clone();
+        for (key, value) in &other.trie {
+            if let Some(existing) = res.trie.get(key) {
+                res.trie.insert(key.clone(), (*existing).union(value));
+            } else {
+                res.trie.insert(key.clone(), value.clone());
+            }
```

Comment:
here and elsewhere: the `.entry()` API will avoid the double hashmap lookup.  (In this code, you're doing a hashmap lookup for `.get()` and then repeating the lookup for the same key for `.insert()`)